go/coinstacks/binance/websocket.go

```diff
@@ -0,0 +1,159 @@
+package binance
+
+import (
+	"context"
+	"net"
+	"net/url"
+	"strconv"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/shapeshift/unchained/pkg/cosmos"
+	tmrand "github.com/tendermint/tendermint/libs/rand"
+	tmsync "github.com/tendermint/tendermint/libs/sync"
+	coretypes "github.com/tendermint/tendermint/rpc/core/types"
+	rpctypes "github.com/tendermint/tendermint/rpc/jsonrpc/types"
+	"github.com/tendermint/tendermint/types"
+	"gitlab.com/thorchain/binance-sdk/client/rpc"
+)
+
+// WSClient wraps the binance-sdk WSClient to properly implement the WebsocketHandler interface
+type WSClient struct {
+	*rpc.WSClient
+	maxReconnectAttempts int
+	mtx                  tmsync.RWMutex
+	nextReqID            int
+	onReconnect          func()
+	reconnecting         bool
+}
+
+func NewWebsocketClient(conf Config, blockService *cosmos.BlockService, errChan chan<- error) (*cosmos.WSClient, error) {
+	wsURL, err := url.Parse(conf.WSURL)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to parse WSURL: %s", conf.WSURL)
+	}
+
+	responsesCh := make(chan rpctypes.RPCResponse)
+
+	client := &WSClient{
+		WSClient: rpc.NewWSClient(wsURL.String(), "/websocket", responsesCh),
+	}
+
+	// use default dialer
+	client.Dialer = net.Dial
+
+	unmarshal := func(bz []byte, result *coretypes.ResultEvent) error {
+		err := conf.Encoding.Amino.UnmarshalJSON(bz, result)
+		if err != nil {
+			return err
+		}
+
+		switch v := result.Data.(type) {
+		case EventDataNewBlockHeader:
+			result.Data = v.ToEventDataNewBlockHeader()
+		}
+
+		return nil
+	}
+
+	ws := cosmos.NewBaseWebsocketClient(blockService, client, conf.Encoding, unmarshal, responsesCh, errChan)
+
+	MaxReconnectAttempts(10)(client)
+	OnReconnect(func() {
+		logger.Info("OnReconnect triggered: resubscribing")
+		_ = client.Subscribe(context.Background(), types.EventQueryTx.String())
+		_ = client.Subscribe(context.Background(), types.EventQueryNewBlockHeader.String())
+	})(client)
+
+	return ws, nil
+}
+
+func MaxReconnectAttempts(max int) func(*WSClient) {
+	return func(c *WSClient) {
+		c.maxReconnectAttempts = max
+	}
+}
+
+func OnReconnect(cb func()) func(*WSClient) {
+	return func(c *WSClient) {
+		c.onReconnect = cb
+	}
+}
+
+func (c *WSClient) Start() error {
+	err := c.WSClient.Start()
+	if err != nil {
+		return err
+	}
+
+	go c.reconnectRoutine()
+
+	return nil
+}
+
+func (c *WSClient) Subscribe(ctx context.Context, query string) error {
+	return c.WSClient.Subscribe(ctx, c.nextRequestID(), query)
+}
+
+func (c *WSClient) UnsubscribeAll(ctx context.Context) error {
+	return c.WSClient.UnsubscribeAll(ctx, c.nextRequestID())
+}
+
+func (c *WSClient) nextRequestID() rpctypes.JSONRPCStringID {
+	c.mtx.Lock()
+	id := c.nextReqID
+	c.nextReqID++
+	c.mtx.Unlock()
+	return rpctypes.JSONRPCStringID(strconv.Itoa(id))
+}
+
+func (c *WSClient) reconnectRoutine() {
+	for range c.Quit() {
```

Comment:
trapping on quit as binance-sdk websocket client does not handle errors gracefully and aborts, need to test this reconnect logic works. it is possible it will result in attempting to close an already closed channel and panic...

Reply:
no longer needed with sdk fork